File: app/schemas/gamificacao.py
"""Schemas de avatar/apelido, pontuacao/sequencia de dias e emblemas."""

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

AvatarCodigo = Literal[
    "raposa", "coruja", "gato", "passaro", "urso", "lobo", "tartaruga", "esquilo"
]


class PerfilJogoRequest(BaseModel):
    apelido: str | None = Field(default=None, max_length=40)
    avatar_codigo: AvatarCodigo | None = None


class PerfilJogoResponse(PerfilJogoRequest):
    aluno_id: uuid.UUID

    model_config = {"from_attributes": True}


class PontuacaoResponse(BaseModel):
    aluno_id: uuid.UUID
    pontos: int
    sequencia_dias: int
    maior_sequencia_dias: int
    ultima_atividade_em: date | None

    model_config = {"from_attributes": True}


class EmblemaResponse(BaseModel):
    id: uuid.UUID
    codigo: str
    nome: str
    descricao: str | None

    model_config = {"from_attributes": True}


class EmblemaConquistadoResponse(EmblemaResponse):
    conquistado_em: datetime

File: frontend/src/lib/api/gamificacao.ts
import { apiClient } from './client'
import type {
  EmblemaConquistadoResponse,
  EmblemaResponse,
  PerfilJogoRequest,
  PerfilJogoResponse,
  PontuacaoResponse,
} from './types'

export async function obterMeuAvatar(): Promise<PerfilJogoResponse> {
  const { data } = await apiClient.get<PerfilJogoResponse>('/me/avatar')
  return data
}

export async function atualizarMeuAvatar(payload: PerfilJogoRequest): Promise<PerfilJogoResponse> {
  const { data } = await apiClient.put<PerfilJogoResponse>('/me/avatar', payload)
  return data
}

export async function obterMinhaPontuacao(): Promise<PontuacaoResponse> {
  const { data } = await apiClient.get<PontuacaoResponse>('/me/pontuacao')
  return data
}

export async function obterPontuacaoDeAluno(alunoId: string): Promise<PontuacaoResponse> {
  const { data } = await apiClient.get<PontuacaoResponse>(`/alunos/${alunoId}/pontuacao`)
  return data
}

export async function listarEmblemas(): Promise<EmblemaResponse[]> {
  const { data } = await apiClient.get<EmblemaResponse[]>('/emblemas')
  return data
}

export async function listarMeusEmblemas(): Promise<EmblemaConquistadoResponse[]> {
  const { data } = await apiClient.get<EmblemaConquistadoResponse[]>('/me/emblemas')
  return data
}
